### PR TITLE
add schema to store national identity information for ~users

### PR DESCRIPTION
**Pull request by whit537**
&#x2B11; #3994 

#4006 &rarr;

Alright, let's do this. The eating of an :elephant: begins with the first bite! :fork_and_knife:

#### Punchlist

 - [x] require `schema_version` in `info` field
 - [x] add client-side encryption to `info` field(!)
 - [x] log events
 - [x] finish docs
 - [x] clean up commits

#### PRs

 - [x] identity preliminaries: #4004 (against `master`), includes ...
  - [x] fix the Makefile for sphinx: #3996 (against `master`)
  - [x] make way for refactoring participant: #3997 (against `master`)
  - [x] document `add_event`
  - [x] prune dead `crypto` code
 - [x] add new crypto code: #3998 (against `master`)
 - [x] add schema, data, and Python for countries: #3999 (against #3998)
 - [x] add schema and Python for storing and retrieving identities: #4000 (against #3999)
 - [x] add Python for verifying and unverifying identities: #4001 (against #4000)
 - [x] add Python for clearing an identity: #4002 (against #4001)
 - [x] add schema and trigger for `participants.has_verified_identity`: #4003 (against #4002)

### CODE REVIEW

**Comment by whit537** [comment]
I'm thinking that we'll use [client-side encryption](http://www.postgresql.org/docs/9.4/static/encryption-options.html) on the `personal_identities.info` field.

At this point we know very little about what exactly we'll need to collect. It may vary from country to country, and I suspect it'll be driven largely by the third-party verification vendor(s) we end up going with. Going with a JSON field for now seemed like the easiest way to get started.

I think we should allow multiple identities per ~user, but only one per ~user per country (we should account for dual citizenship, in other words).

I guess let's log changes to the `events` table, with the old info in the payload as with ... wherever else it was that we did that recently. :-)

**Comment by whit537** [comment]
I'm expecting that eventually we'll add a `organizational_identities` table that links with `teams`.

**Comment by whit537** [comment]
> At a minimum, the bank must obtain the following identifying information from each customer before opening the account:
>
Name.
Date of birth for individuals.
Address.
Identification number.

https://github.com/gratipay/gratipay.com/issues/3289#issuecomment-107086274

**Comment by whit537** [comment]
Turns out we still have an identity form still [hanging around](https://github.com/gratipay/gratipay.com/blob/035e5b755344e5649686a0535710ea68c1020d89/www/~/%25username/identity.spt) post-Balanced.

**Comment by mattbk** [comment]
> I guess let's log changes to the events table, with the old info in the payload as with ... wherever else it was that we did that recently. :-)

https://github.com/gratipay/gratipay.com/pull/3923#issuecomment-187322860

File: sql/countries.sql [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+INSERT INTO countries (name, code) VALUES ('Vanuatu', 'VUT');
+INSERT INTO countries (name, code) VALUES ('Venezuela (Bolivarian Republic of)', 'VEN');
+INSERT INTO countries (name, code) VALUES ('Viet Nam', 'VNM');
+INSERT INTO countries (name, code) VALUES ('Virgin Islands (British)', 'VGB');
+INSERT INTO countries (name, code) VALUES ('Virgin Islands (U.S.)', 'VIR');
+INSERT INTO countries (name, code) VALUES ('Wallis and Futuna', 'WLF');
+INSERT INTO countries (name, code) VALUES ('Western Sahara*', 'ESH');
+INSERT INTO countries (name, code) VALUES ('Yemen', 'YEM');
+INSERT INTO countries (name, code) VALUES ('Zambia', 'ZMB');
+INSERT INTO countries (name, code) VALUES ('Zimbabwe', 'ZWE');
```
Comment (aandis) [inline sql/countries.sql]: geez. Why not a single multi row `INSERT` statement?

Comment (whit537) [inline sql/countries.sql]: Laziness, sorry. Cleaned up in d78a132e9b03392e26b504ec54714ec2fea1b56f. :blush:

**Comment by whit537** [comment]
Rebased on master at 4270aa0. Previous head was 1b97231aa7dff1f6c4232c8c814d1968caec5b33.

**Comment by whit537** [comment]
Lotta discussion over here. A summary:

 - Individuals can renounce their citizenship. We should allow ~users to deactivate the national identities they have stored with us.
 - Countries fork and merge. This would be case of mass deactivation and reregistration.
 - Deactivating one's last `participant_identities` should a) either remove you from all Teams, or b) be prevented if you're still on a Team. Probably the former?
 - Our recent pattern has been to include old info in the `events` table when we log changes. However, we want to keep PII in one place to, we don't want to also store it in the `events` table. What we're thinking is that we'll use our old ctime/mtime pattern (a la `payment_instructions`)—still also logging to `events`, but with the ids of the old and new records rather than the full info.
 - We don't want to overengineer the `info` schema right now. We'll collect (https://github.com/gratipay/gratipay.com/pull/3976#issuecomment-204204890) name, address, date of birth, and identity number (allowing for national id, drivers license, and/or passport). We should version the schema so we're in a better position as we need to start validating it more strictly.

**Comment by whit537** [comment]
MIT flags: https://www.gosquared.com/resources/flag-icons/

**Comment by kaguillera** [comment]
Sprite/css flags alternative as discussed
https://www.flag-sprites.com/

**Comment by kaguillera** [comment]
flags

CSS: (https://gist.github.com/kaguillera/267ee747bd1e7679732b81d7b2f78d10)

**Comment by whit537** [comment]
I'm now questioning why we need to retain old PII. Seems safer to simply delete or overwrite it as requested.

**Comment by whit537** [comment]
More like that (previous was fd6d63599cf657417ba28ad010222fe7279674db).

**Comment by whit537** [comment]
screen shot 2016-04-28 at 11 47 44 pm

**Comment by whit537** [comment]
Okay! Got a good chunk of work done here. I am ruminating on how to carve this up into easily reviewable PRs ...

**Comment by whit537** [comment]
Alright, got some good commits and I made PRs for them. I've update the ticket description here with a list. It's a little wonky because half-way through I started making the PRs against the previous PR instead of `master` because I thought that might be easier, but due to the way we have Travis configured we're not getting CI results unless the PR is against `master`. I've updated this PR to contain all of the commits, and I've pushed them one at a time so we can see Travis results for them.

**Comment by whit537** [comment]
Hmm ... actually, it looks like Travis may still be confused.

**Comment by whit537** [comment]
Okay, I've gone ahead and consolidated the first four commits into a PR, #4004. I think we should start with that as a piece for review and merge, and then we can tackle the other six. I guess you can start reviewing the six either in the commits on this PR or in the separate PRs (again, see the description on this ticket for links). I think it probably makes sense to rebase and merge in ascending order, rather starting at the end and merging back up the chain. In other words, I guess we'll close instead of merging the latter five PRs, making *new* PRs against `master` for each commit as we're ready.

Of course, you may find enough wrong with this code that we'll end up with a new commit history anyway. :-)

**Comment by whit537** [comment]
Okay! I've marked this one and #4004 as ready for review! Here we go! 🐘 🍴💃 

cc: @rohitpaulk @aandis @rorepo @kaguillera @clone1018 et al.

**Comment by whit537** [comment]
We shouldn't allow users to store a national identity without a verified email address.

**Comment by whit537** [comment]
Rebased on master @279b7ca, previous head was 385fae2498842f1c5301f755858ba40d0aed247c.

**Comment by whit537** [comment]
On to #3998! 😮

**Comment by whit537** [comment]
> We shouldn't allow users to store a national identity without a verified email address.

For some reason I thought we already had `participants.has_verified_email`, a la `has_verified_identity` in #4003, but now I don't see it, either in Postgres or in Python. I guess we should insert a PR into the chain that has that, and then use it subsequently. I suppose it should precede https://github.com/gratipay/gratipay.com/pull/4000, because we want to make storage of an identity conditional on it.

**Comment by whit537** [comment]
> I thought we already had `participants.has_verified_email`

It's not a boolean, it's the address itself: `participants.email_address`.

**Comment by whit537** [comment]
Old head was 9907401d90528f061df0928cf61828963550db60.

**Pull request closed without merging**